Check both browser and chrome objects in enumerateFramesAndNamespaceObjects

b-weinstein opened the pull request:
#### 5e8c33f6e1ce4b6f4f7d3fadbdd34ea40b2ce8f5
<pre>
Check both browser and chrome objects in enumerateFramesAndNamespaceObjects
<a href="https://bugs.webkit.org/show_bug.cgi?id=270657">https://bugs.webkit.org/show_bug.cgi?id=270657</a>
<a href="https://rdar.apple.com/123409359">rdar://123409359</a>

Reviewed by Timothy Hatcher.

Some extensions use a polyfill to overwrite the browser object with their own proxy. The Blue Canoe extension was doing this,
and it led to WebExtensionContextProxy::enumerateFramesAndNamespaceObjects not being able to find the namespace object for the extension,
since we were only checking the `browser` object.

To fix this, check both `browser` and `chrome`, and use whichever one is valid.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndNamespaceObjects):

Canonical link: <a href="https://commits.webkit.org/275809@main">https://commits.webkit.org/275809@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b9d4f60c095a529a2b9177cc3bb483086c355593

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21929 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45309 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45526 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39033 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25630 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19309 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18954 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36944 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16481 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16562 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/964 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/39084 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38336 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/47049 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17735 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14622 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/42258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/19360 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40913 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9565 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19524 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/18990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
```diff
@@ -189,15 +189,21 @@ void WebExtensionContextProxy::enumerateFramesAndNamespaceObjects(const Function
+        if (namespaceObject && JSValueIsObject(context, namespaceObject)) {
+            if (RefPtr namespaceObjectImpl = toWebExtensionAPINamespace(context, namespaceObject)) {
+                function(frame, *namespaceObjectImpl);
+                continue;
```
This is wrong, when it finds it, it should use it. This will skip the whole frame.

You want something like this:

```c++
RefPtr namespaceObjectImpl;

auto browserNamespaceObject = JSObjectGetProperty(context, globalObject, toJSString("browser").get(), nullptr);
if (browserNamespaceObject && JSValueIsObject(context, browserNamespaceObject))
    namespaceObjectImpl = toWebExtensionAPINamespace(context, browserNamespaceObject);

if (!namespaceObjectImpl) {
    auto chromeNamespaceObject = JSObjectGetProperty(context, globalObject, toJSString("chrome").get(), nullptr);
    if (chromeNamespaceObject && JSValueIsObject(context, chromeNamespaceObject))
        namespaceObjectImpl = toWebExtensionAPINamespace(context, chromeNamespaceObject);
}

if (!namespaceObjectImpl)
    continue;
```

b-weinstein replied:
Line 196 calls the function, so it does get used, but I can structure it like this if you prefer.

xeenon replied:
Yep missed that!

b-weinstein committed:
Check both browser and chrome objects in enumerateFramesAndNamespaceObjects
https://bugs.webkit.org/show_bug.cgi?id=270657
rdar://123409359

Reviewed by Timothy Hatcher.

Some extensions use a polyfill to overwrite the browser object with their own proxy. The Blue Canoe extension was doing this,
and it led to WebExtensionContextProxy::enumerateFramesAndNamespaceObjects not being able to find the namespace object for the extension,
since we were only checking the `browser` object.

To fix this, check both `browser` and `chrome`, and use whichever one is valid.

* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.cpp:
(WebKit::WebExtensionContextProxy::enumerateFramesAndNamespaceObjects):

Canonical link: https://commits.webkit.org/275809@main
